### PR TITLE
Rework of TwoLevelTree

### PR DESCRIPTION
**Pull request by JulStraus**
This PR focuses on a rework of the `TwoLevelTree` structure. While it is not yet finalized, it includes the following main points:

1. Reworked the structures based on the changes in the iteration procedures.
2. Included support for proper strategic periods analyses, operational scenarios, and representative periods. The combination of operational scenarios and representative periods is not yet included.
3. Reworked the `StrategicStochasticProfiles` to be consistent with the other type of profiles.
4. `WithPrev` on the strategic periods of a `TwoLevelTree` is now basing it on the parent node. That is no changes are required 

There are still a few things to be finalized:

- [x] Include support for the combination of operational scenarios and representative periods. This should be rather straightforward.
- [ ] Rethink constructors and the top type, _i.e._, `TwoLevelTree`.
- [x] Improve on testing.
- [ ] Think more about parametric types for the new created types.
- [ ] Think about what to export

Minor comments from my end:

- I am not fully satisfied with the `TwoLevelTree` at the moment, as it includes all nodes from the beginning. This does however work quite well. It could be reworked eventually.
- Branch probability as a field: I added it separately to avoid any issue with the probability of the individual operational scenarios.

As a last point: All tests are running. I included new tests for the combination with `RepresentativePeriods`. However, I cannot guarantee that all functions for other subtypes are working, _e.g._, functions for `StrategicPeriod`. I though of formatting in the end, not before. The key reason for that is that I want to avoid any issues ;)

### CODE REVIEW

**Comment by JulStraus** [comment]
The documentation fails in the current stage as we do not provide a library of internal types/functions. I would add these eventually.

File: src/profiles.jl [hunk 1]
```diff
@@ -165,18 +165,52 @@ function Base.getindex(
...
+    [OperationalProfile([21, 22]), OperationalProfile([31, 32])]
+])
+```
+"""
+struct StrategicStochasticProfile{T<:Duration,P<:TimeProfile{T}} <: TimeProfile{T}
+    vals::Vector{<:Vector{P}}
 end
-function Base.getindex(ssp::StrategicStochasticProfile, i::TimePeriod)
-    return ssp.vals[_strat_per(i)][_branch(i)]
+function StrategicStochasticProfile(vals::Vector{<:Vector{<:Duration}})
```
Comment (trulsf) [inline src/profiles.jl:193]: Use `Number` instead of `Duration`

File: src/tree_nodes.jl [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+abstract type AbstractTreeNode{T} <: TimeStructure{T} end
+
+"""
+    AbstractTreeStructure
+
+Abstract base type for all tree timestructures within a [`TwoLevelTree`](@ref) type.
+"""
+abstract type AbstractTreeStructure end
+
+Base.length(oscs::AbstractTreeStructure) = length(_oper_struct(oscs))
```
Comment (trulsf) [inline src/tree_nodes.jl:15]: Maybe use a different variable name than `oscs`?

**Comment by JulStraus** [comment]
I added now a first take on the documentation for compilation properties. There is some weird behavior in `Documenter`, but I cannot say what a solution may be. That is specifically for cross referencing in docstrings.

I personally would say that it is more or less now a version which can be seen as acceptable. I personally would however do changes to it in PR #23 as the readjustments may simplify the overall understandability. Note that this will be a separate commit in said branch in the end.

Similarly, both open points could be an open discussion for us pre releasing the new version.

**Review by trulsf — APPROVED** [review]

**Pull request merged**